src/aap_eda/tasks/ruleset.py

```diff
@@ -29,34 +30,98 @@
+        if (
+            not activation
+            or str(activation.status) == ActivationStatus.DELETING.value
+        ):
+            logger.info(f"Activation id: {activation_id} is deleted")
+            return
 
-    # TODO(hsong): move this to views
-    if is_restart:
-        activation.restart_count += 1
-        activation.save(update_fields=["restart_count", "modified_at"])
-    logger.info(f"Task started: Activate rulesets ({activation.name})")
+        logger.info(
+            f"Activating activation id: {activation_id} requested "
+            f"by {requester}"
+        )
+
+        activation.status = ActivationStatus.STARTING
+        activation.save(update_fields=["status"])
 
     if activation.is_enabled:
+        # Note: expect scheduler to set requester as "SCHEDULER"
+        if requester == "SCHEDULER":
+            activation.restart_count += 1
+            activation.save(update_fields=["restart_count", "modified_at"])
+
         ActivateRulesets().activate(activation)
 
-        logger.info(f"Task finished: Rulesets ({activation.name}).")
+        logger.info(
+            f"Task finished: activation {activation.name}"
+            f" {activation.status}."
+        )
     else:
         logger.info(
-            f"Task finished: Rulesets ({activation.name}) has been disabled."
+            f"Task finished: activation id {activation_id} has been disabled."
         )
 
 
 @job("default")
-def deactivate(activation_id: int, is_delete: bool = False) -> None:
-    activation = models.Activation.objects.get(id=activation_id)
+def deactivate(
+    activation_id: int, requester: str = "User", delete: bool = False
+) -> None:
+    logger.info(
+        f"Disabling activation id: {activation_id} requested by {requester}"
+    )
+
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if not activation:
+            logger.warning(f"Activation {activation_id} is deleted.")
+            return
+
+        _perform_deactivate(activation)
+
+        if delete:
+            logger.info(f"Activation {activation.name} is deleted")
+            activation.delete()
+        else:
+            activation.current_job_id = None
+            activation.status = ActivationStatus.STOPPED
+            activation.save()
+
+
+@job("default")
+def restart(activation_id: int, requester: str = "User") -> None:
+    logger.info(
+        f"Restarting activation id: {activation_id} requested by {requester}"
+    )
+
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if not activation:
+            logger.warning(f"Activation {activation_id} is deleted.")
+            return
+
+        _perform_deactivate(activation)
+
+        queue = get_queue(name="activation")
```

Comment:
Before this line we should validate the activation is restartable, otherwise don't bother to add it to the queue. It make sense to change the status to pending before queuing so the activate function only need to validate the activation is in pending status.